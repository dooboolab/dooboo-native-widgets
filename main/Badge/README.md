# Badge  
> [Badge] component that can be used inside product. Has `OpacityVisible` state, `badgeZero` state.  
<br/>
<br/>

## Preview 

| Badge(default) | Badge(color, count, showZero) |
| :---------------: | :--------------: |
| <img src="https://user-images.githubusercontent.com/37579661/92614370-eff66380-f2f6-11ea-9f46-35f5cc871a72.png" width= "50%"/> | <img src="https://user-images.githubusercontent.com/37579661/92614395-f553ae00-f2f6-11ea-9663-f507d62c0a7c.png" width= "50%"/> |

| Badge(maximumCount, opacityVisible) | Badge(border, textColor) | Badge(variant, position) |
| :---------------: | :--------------: | :-------------: |
| <img src="https://user-images.githubusercontent.com/37579661/92614433-01d80680-f2f7-11ea-99a5-ed89ff14216c.png" width= "100%"/> | <img src="https://user-images.githubusercontent.com/37579661/92614441-056b8d80-f2f7-11ea-9e40-0ca17b6db965.png" width= "100%"/> |  <img src="https://user-images.githubusercontent.com/37579661/92616226-ee2d9f80-f2f8-11ea-9573-7ebec6f36ce6.png" width= "100%"/> |



## Props  
  
|                       | required | types                 | default   |description |
| :-------------------: | -------- | :-------------------: | :-------: | ---------- |
| **count**             |          | number                |    1      | Proportionally controls opacity of badge color |
| **color**             |          | string                | '34AFF9'  | Set badge color |
| **maximumValue**      |          | number                |     300   | Set maximum limit where opacity stay 1.0, and if count exceeds maximumValue, badge color opacity sets to 0.6 | 
| **showZero**          |          | boolean               |    true   | Set whether to show '0' count number in badge or not |
| **opacityVisible**    |          | boolean               |    true   | Set application of change in color opacity |
| **variant**           |          | string                |'standard' | Set shape of the badge. Available choices are 'dot' and 'standard' |
| **badgePlacement**          |          | string                |  'right'  | Set position of the badge. Available choices are 'left', 'right' , 'top', 'down'. |
| **border**            |          | string                |           | Set border color of the badge. |
| **textColor**         |          | string                | '#FFFFFF' | Set textColor of the count number inside the badge. |


<br/>
<br/>  

## Installation  
  
```sh
yarn add dooboo-ui
```
<br/>
<br/>


## Getting started  
  
- Import

  ```javascript
  import { Badge } from 'dooboo-ui';
  ```  
<br/>
<br/>


## Usage

```tsx

const Container = styled.View`
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

const ShowContainer = styled.View`
  justify-content: center;
  align-items: center;
  flex-direction: row;
`;

const StyledTitle = styled.Text`
  font-size: 20px;
  margin-bottom: 20px;
`;

const StyledView: any = styled.View`
  margin: 20px;
  width: 100px;
  height: 100px;
  background-color: #BFBFBF
  border-radius: 12;
  box-shadow: 3px 3px 2px rgba(0, 0, 0, 0.5);
  elevation: 1;
`;

const Badge1 = (): React.ReactElement => {
  return (
    <SafeAreaView>
      <ScrollView
        contentContainerStyle={{
          alignSelf: 'stretch',
          paddingHorizontal: 20,
          paddingVertical: 30,
        }}>
        <Container>
          <View>
            <StyledTitle>Badge (default)</StyledTitle>
            <StyledView>
              <Badge />
            </StyledView>
          </View>

          <View style={{ marginTop: 20 }}>
            <Container>
              <StyledTitle>Badge (color, count, showZero)</StyledTitle>
            </Container>
            <ShowContainer>
              <StyledView>
                <Badge color="#6B98F2" count={17} />
              </StyledView>
              <StyledView>
                <Badge color="#6B98F2" count={0} showZero />
              </StyledView>
            </ShowContainer>
          </View>

          <View style={{ marginTop: 20 }}>
            <Container>
              <StyledTitle>Badge (variant, position)</StyledTitle>
            </Container>
            <ShowContainer>
              <StyledView>
                <Badge variant="dot" color="#ED2939" />
              </StyledView>
              <StyledView>
                <Badge color="#F2B9AC" count={100} position="left" />
              </StyledView>
            </ShowContainer>
          </View>

          <View style={{ marginTop: 20 }} >
            <Container>
              <StyledTitle>Badge (maximumCount, opacityVisible)</StyledTitle>
            </Container>
            <ShowContainer>
              <StyledView>
                <Badge
                  opacityVisible={false}
                  color="#34AFF9"
                  count={300}
                  maximumCount={100}
                />
              </StyledView>
              <StyledView>
                <Badge
                  opacityVisible={true}
                  color="#34AFF9"
                  count={300}
                  maximumCount={100}
                />
              </StyledView>
            </ShowContainer>
          </View>

          <View style={{ marginTop: 20 }}>
            <Container>
              <StyledTitle>Badge (border, textColor)</StyledTitle>
            </Container>
            <ShowContainer>
              <StyledView>
                <Badge color="#91AAF2" border="#F2B9AC" count={37} />
              </StyledView>
              <StyledView>
                <Badge color="white" border="#F2B9AC" textColor="#F2B9AC" count={50} />
              </StyledView>
            </ShowContainer>
          </View>
        </Container>
      </ScrollView>
    </SafeAreaView>
  );
};


```