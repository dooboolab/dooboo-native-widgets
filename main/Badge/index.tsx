import React, { FC } from 'react';
import { StyleSheet } from 'react-native';
import styled from 'styled-components/native';

interface StyleProps {
  color?: string;
  opacity?: number;
  badgePlacement?: 'left' | 'top' | 'bottom' | 'right';
  border?: string;
  textColor?: string;
}

export interface BadgeProps extends StyleProps{
  count?: number;
  maximumCount?: number;
  showZero?: boolean;
  opacityVisible?: boolean;
  variant?: string;
}

const StyledView = styled.View<StyleProps>`
  position: absolute;
  top: -15px;
  ${(props) => props.badgePlacement}: -10px;
  width: auto;
  min-width: 45px;
  height: 45px;
  border-color: ${(props) =>
    props.border ? props.border : '#00ff0000'};
  border-width: 3px;
  background-color: ${(props) => props.color};
  border-radius: 50px;
  justify-content: center;
  align-items: center;
  opacity: ${(props) => props.opacity};
`;

const StyledText = styled.Text<StyleProps>`
  color: ${(props) => props.textColor};
  text-align: center;
  padding: 5px;
  margin-left: 3px;
  margin-right: 3px;
`;

const StyledDotView = styled.View<StyleProps>`
  position: absolute;
  top: -5px;
  ${(props) => props.badgePlacement}: -5px;
  width: 20px;
  height: 20px;
  background-color: ${(props) => props.color};
  border-radius: 50px;
  justify-content: center;
  align-items: center;
`;

const styles = StyleSheet.create({
  fontDesign: {
    fontWeight: '500',
    fontSize: 20,
  },
});

const Badge: FC<BadgeProps> = (props) => {
  const {
    count = 1,
    color = '#34AFF9',
    maximumCount = 300,
    showZero,
    opacityVisible = true,
    variant = 'standard',
    badgePlacement: position = 'right',
    border,
    textColor = '#FFFFFF',
  } = props;

  if (!showZero) {
    if (count === 0) return null;
  }

  switch (true) {
    case variant === 'dot':
      return <StyledDotView color={color} badgePlacement={position} />;

    case maximumCount >= count:
      return (
        <StyledView
          color={color}
          opacity={1}
          badgePlacement={position}
          border={border}>
          <StyledText textColor={textColor} style={styles.fontDesign}>
            {count}
          </StyledText>
        </StyledView>
      );

    case maximumCount < count && opacityVisible:
      return (
        <StyledView
          color={color}
          opacity={0.6}
          badgePlacement={position}
          border={border}>
          <StyledText textColor={textColor} style={styles.fontDesign}>
            {count + '+'}
          </StyledText>
        </StyledView>
      );

    case maximumCount < count && !opacityVisible:
      return (
        <StyledView
          color={color}
          opacity={1}
          badgePlacement={position}
          border={border}>
          <StyledText textColor={textColor} style={styles.fontDesign}>
            {count + '+'}
          </StyledText>
        </StyledView>
      );

    default:
      return (
        <StyledView
          color={color}
          opacity={1}
          badgePlacement={position}
          border={border}>
          <StyledText textColor={textColor} style={styles.fontDesign}>
            {count}
          </StyledText>
        </StyledView>
      );
  }
};

export { Badge };
