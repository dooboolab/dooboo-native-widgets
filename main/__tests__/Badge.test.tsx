import { Badge, BadgeProps } from '../Badge';
import React, { ReactElement } from 'react';
import { RenderResult, render } from '@testing-library/react-native';

let props: BadgeProps;
let component: ReactElement;
let testingLib: RenderResult;

const createTestProps = (
  obj?: Record<string, unknown>,
): Record<string, unknown> => ({
  ...obj,
});

describe('[Badge] render', () => {
  it('should render without crashing', () => {
    props = createTestProps();
    component = <Badge {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
    expect(testingLib.baseElement).toBeTruthy();
  });

  it('should render with + when count is over than max count', () => {
    props = createTestProps({
      maximumCount: 300,
      count: 500,
    });

    component = <Badge {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
    expect(testingLib.baseElement).toBeTruthy();
  });

  it('should not render opacity when count is over than max count', () => {
    props = createTestProps({
      maximumCount: 300,
      count: 500,
      opacityVisible: false,
    });

    component = <Badge {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
    expect(testingLib.baseElement).toBeTruthy();
  });

  it('should render dot badge', () => {
    props = createTestProps({
      variant: 'dot',
    });

    component = <Badge {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
    expect(testingLib.baseElement).toBeTruthy();
  });

  it('should render left badge', () => {
    props = createTestProps({
      position: 'left',
    });

    component = <Badge {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
    expect(testingLib.baseElement).toBeTruthy();
  });

  it('should render badge when wrong variant', () => {
    props = createTestProps({
      variant: 'dott',
    });

    component = <Badge {...props} />;
    testingLib = render(component);

    expect(testingLib.baseElement).toMatchSnapshot();
    expect(testingLib.baseElement).toBeTruthy();
  });
});
