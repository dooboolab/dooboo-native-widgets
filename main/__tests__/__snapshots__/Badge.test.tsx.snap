// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Badge] render should not render opacity when count is over than max count 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      badgePlacement="right"
      color="#34AFF9"
      opacity={1}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#34AFF9",
            "borderColor": "#00ff0000",
            "borderRadius": 50,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "minWidth": 45,
            "opacity": 1,
            "position": "absolute",
            "right": -10,
            "top": -15,
            "width": "auto",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "marginLeft": 3,
              "marginRight": 3,
              "paddingBottom": 5,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 5,
              "textAlign": "center",
            },
            Object {
              "fontSize": 20,
              "fontWeight": "500",
            },
          ]
        }
        textColor="#FFFFFF"
      >
        500+
      </Text>
    </View>
  </View>
</View>
`;

exports[`[Badge] render should render badge when wrong variant 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      badgePlacement="right"
      color="#34AFF9"
      opacity={1}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#34AFF9",
            "borderColor": "#00ff0000",
            "borderRadius": 50,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "minWidth": 45,
            "opacity": 1,
            "position": "absolute",
            "right": -10,
            "top": -15,
            "width": "auto",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "marginLeft": 3,
              "marginRight": 3,
              "paddingBottom": 5,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 5,
              "textAlign": "center",
            },
            Object {
              "fontSize": 20,
              "fontWeight": "500",
            },
          ]
        }
        textColor="#FFFFFF"
      >
        1
      </Text>
    </View>
  </View>
</View>
`;

exports[`[Badge] render should render dot badge 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      badgePlacement="right"
      color="#34AFF9"
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#34AFF9",
            "borderRadius": 50,
            "height": 20,
            "justifyContent": "center",
            "position": "absolute",
            "right": -5,
            "top": -5,
            "width": 20,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`[Badge] render should render left badge 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      badgePlacement="right"
      color="#34AFF9"
      opacity={1}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#34AFF9",
            "borderColor": "#00ff0000",
            "borderRadius": 50,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "minWidth": 45,
            "opacity": 1,
            "position": "absolute",
            "right": -10,
            "top": -15,
            "width": "auto",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "marginLeft": 3,
              "marginRight": 3,
              "paddingBottom": 5,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 5,
              "textAlign": "center",
            },
            Object {
              "fontSize": 20,
              "fontWeight": "500",
            },
          ]
        }
        textColor="#FFFFFF"
      >
        1
      </Text>
    </View>
  </View>
</View>
`;

exports[`[Badge] render should render with + when count is over than max count 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      badgePlacement="right"
      color="#34AFF9"
      opacity={0.6}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#34AFF9",
            "borderColor": "#00ff0000",
            "borderRadius": 50,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "minWidth": 45,
            "opacity": 0.6,
            "position": "absolute",
            "right": -10,
            "top": -15,
            "width": "auto",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "marginLeft": 3,
              "marginRight": 3,
              "paddingBottom": 5,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 5,
              "textAlign": "center",
            },
            Object {
              "fontSize": 20,
              "fontWeight": "500",
            },
          ]
        }
        textColor="#FFFFFF"
      >
        500+
      </Text>
    </View>
  </View>
</View>
`;

exports[`[Badge] render should render without crashing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      badgePlacement="right"
      color="#34AFF9"
      opacity={1}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#34AFF9",
            "borderColor": "#00ff0000",
            "borderRadius": 50,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "minWidth": 45,
            "opacity": 1,
            "position": "absolute",
            "right": -10,
            "top": -15,
            "width": "auto",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "marginLeft": 3,
              "marginRight": 3,
              "paddingBottom": 5,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 5,
              "textAlign": "center",
            },
            Object {
              "fontSize": 20,
              "fontWeight": "500",
            },
          ]
        }
        textColor="#FFFFFF"
      >
        1
      </Text>
    </View>
  </View>
</View>
`;
